src/AbstractBaseMap.php

```diff
@@ -0,0 +1,245 @@
+
+    /* Ability retrieve data from an internal container.
+     *
+     * @since [*next-version*]
+     */
+    use GetDataCapableTrait;
+
+    /* Ability to check for data in an internal container.
+     *
+     * @since [*next-version*]
+     */
+    use HasDataCapableTrait;
+
+    /* Ability to retrieve data for a key from a container.
+     *
+     * @since [*next-version*]
+     */
+    use ContainerGetCapableTrait;
+
+    /* Ability to check for data key on a container.
+     *
+     * @since [*next-version*]
+     */
+    use ContainerHasCapableTrait;
+
+    /* Data store awareness
+     *
+     * @since [*next-version*]
+     */
+    use DataStoreAwareContainerTrait;
+
+    /* Factory of iterations.
+     *
+     * @since [*next-version*]
+     */
+    use CreateIterationCapableTrait;
+
+    /* Factory of Iterator exceptions.
+     *
+     * @since [*next-version*]
+     */
+    use CreateIteratorExceptionCapableTrait;
+
+    /* Factory of Out of Range exceptions.
+     *
+     * @since [*next-version*]
+     */
+    use CreateOutOfRangeExceptionCapableTrait;
+
+    /* Factory of Invalid Argument exceptions.
+     *
+     * @since [*next-version*]
+     */
+    use CreateInvalidArgumentExceptionCapableTrait;
+
+    /* Ability to normalize containers.
+     *
+     * @since [*next-version*]
+     */
+    use NormalizeContainerCapableTrait;
+
+    /* Ability to normalize integers.
+     *
+     * @since [*next-version*]
+     */
+    use NormalizeIntCapableTrait;
+
+    /* Ability to normalize strings.
+     *
+     * @since [*next-version*]
+     */
+    use NormalizeStringCapableTrait;
+
+    /* Ability to normalize container keys.
+     *
+     * @since [*next-version*]
+     */
+    use NormalizeKeyCapableTrait;
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public function key()
+    {
+        return $this->_key();
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public function current()
+    {
+        return $this->_value();
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public function next()
+    {
+        $this->_next();
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public function valid()
+    {
+        return $this->_valid();
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public function rewind()
+    {
+        $this->_reset();
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public function getIteration()
+    {
+        return $this->_getIteration();
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    protected function _loop()
+    {
+        $iterator = $this->_getIterator();
+
+        $iterator->next();
+        $iteration = $this->_createIteration($iterator->key(), $iterator->current());
```

Comment:
Check if the `$iterator` is still valid after advancing. If not, the key and value of the created iteration should be `null`.

This change would make this package compatible with PHP 5.4.0, otherwise it is only compatible with PHP 5.4.6 or later.